AudiusProject/audius-protocol
# File storage directory nesting

### Trello Card Link
https://trello.com/c/dyViGxGG/1654-track-upload-errors

### Description
The /file_storage directory is susceptible to hash collisions in ext4 file systems using the half md5 hashing algorithm because we write all files to one directory. Moving to a nested directory model should alleviate these concerns.

### Services

- [ ] Discovery Provider
- [X] Creator Node
- [ ] Identity Service
- [ ] Libs
- [ ] Service Commands
- [ ] Mad Dog

### Does it touch a critical flow like Discovery indexing, Creator Node track upload, Creator Node gateway, or Creator Node file system?
Delete an option.
- 🚨 Yes, this touches how files are stored in the file system


### How Has This Been Tested?

1. Created an account, updated my profile picture and cover photo, uploaded a track with the old file storage system. Fetch to metadata and track segment via the ipfs gateway. Fetch a dir image via the ipfs gateway
2. Applied the new file storage system and uploaded more tracks and updated my profile picture and cover photo again. Fetch to metadata and track segment via the ipfs gateway. Fetch a dir image via the ipfs gateway
3. Used the dapp to play the tracks, view the track cover art, my profile picture and cover photo
4. Make sure a sync was triggered and verify clock state for both nodes was the same in the db
5. Verify the number of files in both volumes were the same and the directory structures were identical by listing out all the files, sorting and doing running the diff command
6. Disable fetching through IPFS for sync, only other cnode gateways
7. Upload another track to verify that sync triggered and fetched all files via the gateways
8. Randomly sample and fetch metadata and track segments via the ipfs gateway
9. Randomly sample and fetch dir images via the ipfs gateway
10. Swap the primary between the two nodes, upload another track and hit the ipfs gateway to fetch the content again
